motionplan/planManager.go

```diff
@@ -107,244 +114,263 @@ func (pm *planManager) PlanSingleWaypoint(ctx context.Context,
 	}
 	opts = append(opts, opt)
 
-	resultSlices, err := pm.planMotion(ctx, goals, seed, opts, nil, 0)
+	resultSlices, err := pm.planAtomicWaypoints(ctx, goals, seed, opts)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to plan path for valid goal: %w", err)
 	}
 	return resultSlices, nil
 }
 
-// planMotion will plan a single motion, which may be composed of one or more waypoints. Waypoints are here used to begin planning the next
-// motion as soon as its starting point is known.
-func (pm *planManager) planMotion(
+// planAtomicWaypoints will plan a single motion, which may be composed of one or more waypoints. Waypoints are here used to begin planning
+// the next motion as soon as its starting point is known. This is responsible for repeatedly calling planSingleAtomicWaypoint for each
+// intermediate waypoint. Waypoints here refer to points that the software has generated to.
+func (pm *planManager) planAtomicWaypoints(
```

Comment:
Not sure what you mean by `AtomicWaypoint` is there maybe another name we could be using for these?

Reply:
"Atomic" means "non-splittable", and is probably the most accurate/descriptive term here to distinguish from user-supplied waypoints, which *are* splittable.

Reply:
Do you have suggestions for alternatives here?

Reply:
Maybe it's fine then.  I guess this is coming from a place of getting gradually more overwhelmed with the code which ends up calling our planners.  We now have as a call stack:

PlanMotion -> motionPlanInternal -> PlanSingleWaypoint -> planAtomicWaypoints -> planParallelRRTMotion -> planAtomicWaypoint -> rrtBackgroundRunner.  I understand the motivation for how we ended up here, but its very confusing (even to me who probably has the next most context here) why the stack is ordered this way and what some of these functions do (and even tracing this out took like 5 minutes).  I guess I'm wondering if maybe finding a better name could help here but maybe the issue is deeper than that.  

I think a think that we could be doing that might alleviate this would be mapping out what the relationship is between these functions in a block diagram

Reply:
Good call, I will make the block diagram.

The call stack in your comment is out of order. It is:

[PlanMotion or equivalent] -> motionPlanInternal -> PlanSingleWaypoint -> planAtomicWaypoints -> planSingleAtomicWaypoint -> planParallelRRTMotion -> rrtBackgroundRunner

PlanSingleWaypoint and planAtomicWaypoints could be pretty trivially combined, but One Giant Function is usually more difficult to read than calling a smaller sub-function with a name describing what it does.

Reply:
Cool thanks.  Maybe we could include it in a readme for the motionplan package.  

I think I'm warming up to `AtomicWaypoint`.  The only other thing I could think of is `SubWaypoint`

Reply:
I like `AtomicWaypoint`, it feels about the same same as an atomic operation.